test/Pool.test.js

```diff
@@ -1315,6 +1315,196 @@ describe('Pool', function () {
+      maxWorkers: 1,
+      onCreateWorker: () => {
+        workerCount += 1;
+      }
+    });
+    
+    return pool.exec('asyncTimeout', [],  {
+    })
+    .timeout(200)
+    .catch(function (err) {
+      assert(err instanceof Promise.TimeoutError);
+      let stats = pool.stats();
+      assert(stats.activeTasks === 0);
+    }).then(function() { 
+      return pool.exec(add, [1, 2]) 
+    }).then(function() {
+      assert.strictEqual(workerCount, 1);
+    });
+  });
+
+  it('should not terminate worker if abort listener is defined dedicated worker with Cancellation', function () {
+    var workerCount = 0;
+    var pool = createPool(__dirname + '/workers/cleanup-abort.js', {
+      maxWorkers: 1,
+      onCreateWorker: () => {
+        workerCount += 1;
+      }
+    });
+    
+    let task = pool.exec('asyncTimeout', [],  {});
+
+    // Wrap in a new promise which waits 50ms
+    // in order to allow the function executing in the
+    // worker to 
+    return new Promise(function(resolve) {
+      setTimeout(function() {
+        resolve();
+      }, 50);
+    }).then(function() {
+        return task
+        .cancel()
+        .catch(function (err) {
+          assert(err instanceof Promise.CancellationError);
+          let stats = pool.stats();
+          assert(stats.activeTasks === 0);
+        }).then(function() { 
+          return pool.exec(add, [1, 2]) 
+        }).then(function() {
+          assert.strictEqual(workerCount, 1);
+        });
+    });
+  });
+
+
+  it('should not terminate worker if abort listener is defined inline worker with Timeout', function () {
+    var workerCount = 0;
+    var pool = createPool({
+      onCreateWorker: () => {
+        workerCount += 1;
+      },
+      maxWorkers: 1,
+    });
+    function asyncTimeout() {
+      var me = this;
+      return new Promise(function () {
+        let timeout = setTimeout(function() {
+            resolve();
+        }, 5000); 
+        me.worker.addAbortListener(function () {
+          return new Promise(function (resolve) {
+            clearTimeout(timeout);
+            resolve();
+          });
+        });
+      });
+    }
+    function add(a, b) { }
+    return pool.exec(asyncTimeout, [],  {
+    })
+    .timeout(200)
+    .catch(function(err) {
+      assert(err instanceof Promise.TimeoutError);
+      assert(pool.stats().activeTasks = 0);
+    }).always(function () {
+      return pool.exec(add, [1, 2]).then(function () {
+        assert.strictEqual(workerCount, 1);
+      }); 
+    });
+  });
+
+  it('should not terminate worker if abort listener is defined inline worker with Cancellation', function () {
+    var workerCount = 0;
+    var pool = createPool({
+      onCreateWorker: () => {
+        workerCount += 1;
+      },
+      maxWorkers: 1,
+    });
+    function asyncTimeout() {
+      var me = this;
+      return new Promise(function () {
+        let timeout = setTimeout(function() {
```

Comment:
how can we know for sure that this timeout is not triggered during the test? Maybe let it throw an exception or print "THIS SHOULD NOT BE PRINTED" or something?

Or: create a flag `let aborted = false`, and set that to true inside the `me.worker.addAbortListener`, and at the end verify that it is indeed true?

Reply:
Good call. updating to reject with an error instance instead of resolving with `void`